File: Proyecto final/Problema 2-20230418T031111Z-001/Problema 2/problema2.py
import numpy as np
import scipy.signal as signal
import matplotlib.pyplot as plt
from scipy.io import wavfile
from scipy.optimize import lsq_linear

fs, x = wavfile.read('x2_U017.wav')  # Read x2_U017.wav

# Get the size of the audio vector
lx = len(x)

fs, y = wavfile.read('y2_U017.wav')  # Read y2_U017.wav

# Plot the vectors
n = np.arange(lx)
fig, (ax1, ax2) = plt.subplots(2, 1)
ax1.stem(n, x)
ax1.set_title('X[n]')
ax2.stem(n, y)
ax2.set_title('Y[n]')
plt.show()

# Calculate coefficients
lx = 5000
q = np.arange(lx)
n = np.arange(len(x))

# Solve system of simultaneous equations
c1 = np.zeros(lx-2000)
c2 = np.zeros(lx-2000)
b1 = np.zeros(lx-2000)

for i in range(2001, lx):
    c1[i-2001] = x[i-2001]
    c2[i-2001] = -y[i-751]
    b1[i-2001] = y[i] - x[i]

D = np.vstack((c1, c2)).T
R = np.linalg.lstsq(D, b1, rcond=None)[0]
print("The values for each of the coefficients are:")
print("A=", R[0])
print("B=", R[1])

# Comparison of values for positions between [2001] and [20016]
print("Comparison of values for positions between [2001] and [20016]")
print("First column: Found function Second row: Original function Third row: Noise s[n]")
m = np.zeros(16)
l = np.zeros(16)
s = np.zeros(16)

for j in range(2001, 2017):
    m[j-2001] = x[j] + R[0]*x[j-2000] + R[1]*y[j-750]
    l[j-2001] = y[j]
    s[j-2001] = l[j-2001] - m[j-2001]

M = np.transpose([m])
L = np.transpose([l])
S = np.transpose([s])
Comparacion = np.hstack((M, L, S))

b = [1, R[0]]
a = [1, R[1]]

# Fractional partial coefficients
r, p, k = signal.residuez(b, a)
r = r.reshape((-1,1))
p = p.reshape((-1,1))

HH = np.zeros((lx,), dtype=np.complex128)
for i in range(1,lx,1):
    z = i
    HH[i] = np.sum(r/(1-p*z**(-1))) + np.sum(k*np.array([1, z**(-1)]))

# Spectral analysis
f = 1/10  # Frequency of 0.1Hz
N1 = 30  # Number of samples
N2 = 120

# Spectral analysis for X
X1 = np.abs(np.fft.fft(x, N1))
X2 = np.abs(np.fft.fft(x, N2))

F1x = np.arange(N1-1)/N1
F2x = np.arange(N2-1)/N2

# Spectral analysis for Y
Y1 = np.abs(np.fft.fft(y, N1))
Y2 = np.abs(np.fft.fft(y, N2))

F1y = np.arange(N1-1)/N1
F2y = np.arange(N2-1)/N2

# Grafica de las funciones

fig, (ax1, ax2) = plt.subplots(2, 1)
ax1.stem(n, x)
ax1.set_title('X[n]')
ax2.stem(n, y)
ax2.set_title('Y[n]')
plt.show()

# Espectros para X[n] con N=30 y N=120
F1x = np.linspace(0, 1, 30)
F2x = np.linspace(0, 1, 120)

# Espectros para Y[n] con N=30 y N=120
F1y = np.linspace(0, 1, 30)
F2y = np.linspace(0, 1, 120)

# Parte real e imaginaria de H[z]
b = np.array([1, 2, 3])
a = np.array([1, -0.5, 0.1])
q = np.arange(0, 50)
HH = np.fft.fft(np.concatenate([b, np.zeros(len(q)-len(b))])) / np.fft.fft(np.concatenate([a, np.zeros(len(q)-len(a))]))

# Gráficas de los espectros
fig3, axs3 = plt.subplots(2, 2, figsize=(10, 8))
axs3[0, 0].stem(F1x, X1, '.')
axs3[0, 0].set_title('Espectro para X[n] con N=30')
axs3[0, 1].stem(F2x, X2, '.')
axs3[0, 1].set_title('Espectro para X[n] con N=120')
axs3[1, 0].stem(F1y, Y1, '.')
axs3[1, 0].set_title('Espectro para Y[n] con N=30')
axs3[1, 1].stem(F2y, Y2, '.')
axs3[1, 1].set_title('Espectro para Y[n] con N=120')

# Parte real e imaginaria de H[z]
fig, axs = plt.subplots(2, 1, figsize=(8,8))
axs[0].stem(q, np.real(HH))
axs[0].set_title('Parte real H[z]')
axs[1].stem(q, np.imag(HH))
axs[1].set_title('Parte imaginaria H[z]')
plt.show()

# Respuesta al impulso y al escalón unitario
y = signal.lfilter(b, a, np.arange(lx))

plt.figure(figsize=(12,6))
plt.plot(y)
plt.title('Respuesta al impulso y al escalon unitario')
plt.show()

# Grafica el diagrama de polos y ceros
z, p, _ = signal.tf2zpk(b, a)
plt.figure(6)
plt.scatter(np.real(z), np.imag(z), marker='o', facecolors='none', edgecolors='r')
plt.scatter(np.real(p), np.imag(p), marker='x', color='b')
plt.title('Diagrama de Polos y Ceros')
plt.show()

File: Proyecto final/Problema 2-20230418T031111Z-001/Problema 2/problema2.m
clc
clear
pkg load signal;
pkg load optim;

[x, fs] = audioread('x2_U017.wav'); % se lee el archivo x2_U017.wav

% Almacenado en la actual carpet de trabajo.
% Calculo del tamaño del vector de audio

lx = length(x);

% se lee el archivo y2_U017.wav
% almacenando en la actual carpeta de trabajo.
% calculo del tamaño del vector de audio

[y, fs] = audioread('y2_U017.wav')

% Grafica de los vectores
n=0:lx-1;
figure(1);
subplot(2,1,1);
stem(n,x);
title('X[n]');
subplot(2,1,2);
stem(n,y);
title('Y[n]');

% funcion [b, a] = CalculoCoef2(x,y)
clc
lx=5000;
q=1:lx;
n=1:length(x);

% Determinacion de ssitema de ecuaciones simultaneas para dos variables con mayor numero de ecuaciones
for i=2001:lx-1
  c1(i)=x(i-2000);
  c2(i)=-y(i-750);
  b1(i)=y(i)-x(i);
endfor
C1=transpose(c1);
C2=transpose(c2);
B=transpose(b1);
D=[C1 C2];
lb = [-Inf; -Inf];
ub = [Inf; Inf];
R=lsqlin(D,B, [], [], [], [], lb, ub);
fprintf('El valor para cada uno de los coeficientes es:');
fprintf('\n A= ')
disp(R(1,1));
fprintf(' B= ')
disp(R(2,1));

% Comparacion evalualndo coeficientes ----
fprintf('Comparacion de valores para las posiciones entre [2001] y [20016]\n Primera columna: Funcion encontrada Segunda Fila: Funcion original Tercera Fila: Ruido s[n]');
for j=2001:2016
  m(j-2000)=x(j)+R(1,1)*x(j-2000)+R(2,1)*y(j-750);
  l(j-2000)=y(j);
  s(j-2000)=l(j-2000)-m(j-2000);
endfor
M=transpose(m);
L=transpose(l);
S=transpose(s);
Comparacion = [M L S]

b(1)=1;
b(2)=R(1,1);
a(1)=1;
a(2)=R(2,1);

fprintf('\n Coeficientes de fracciones parciales: \n');
[r, p, k]=residuez(b,a);
r=transpose(r);
p=transpose(p);

b=1:96000;
b=b*0;
a=b;
b(1)=1;
b(2000)=R(1,1);
a(1)=1;
a(750)=R(2,1);
for i=1:lx
  z=i;
  HH(i)=sum(r./(1-p*z^(-1))) + sum(k*[1 ; z^(-1)]);
endfor
HH;

% Analists espectral

f=1/10; % Frecuencia de 0.1Hz
N1=30; % numero de muestras
N2=120;

% Analisis Espectral para X
% Transformadas:
X1=abs(fft(x, N1));
X2=abs(fft(x,N2));

% Rango normalizado para transformadas:
F1x=[(0:N1-1)/N1];
F2x=[(0:N2-1)/N2];

% Analisis espectral para Y
% Transformadas
Y1=abs(fft(y,N1));
Y2=abs(fft(y,N2));

% Rango normalizado para transformadas
F1y=[(0:N1-1)/N1];
F2y=[(0:N2-1)/N2];

% Grafica de funciones

% Funciones X y Y
figure(2);
subplot(2,1,1);
stem(n,x);
title('X[n]');

subplot(2,1,2);
stem(n,y);
title('Y[n]');

% Graficas de los espectros
figure(3)
subplot(2,2,1);
stem(F1x, X1, '.');
title('Espectro para X[n] con N=30');

subplot(2,2,2);
stem(F2x,X2,'.');
title('Espectro para X[n] con N=120');

subplot(2,2,3);
stem(F1y, Y1, '.');
title('Espectro para Y[n] con N=30');

subplot(2,2,4);
stem(F2y, X2, '.');
title('Espectro para Y[n] con N=120');

% Parte real e imaginaria
figure(4)
subplot(2,1,1);
stem(q, real(HH));
title('Parte real H[z]');

subplot(2,1,2);
stem(q, imag(HH));
title('Parte imaginaria H[z]');

% Respuesta en amplitud y frecuencia
figure(5)
y=filter(b,a,q);
plot(q,y);
title('Respuesta al impulso y al escalon unitario');

% Polos y Ceros
figure(6)
zplane(b,a);
title('Diagrama de Polos y Ceros');
